File: Week2/CS/RandomTimer/RandomTimer/Form1.cs
using System.Text;
using System.Timers;

namespace RandomTimer
{
    public partial class Form1 : Form
    {
        private System.Timers.Timer timer;
        private Random rand;
        private DateTime timerStarted;
        private Boolean isDouble = false;

        public Form1()
        {
            InitializeComponent();
            this.timer = new System.Timers.Timer(1000);
            this.timer.AutoReset = true;
            this.timer.Elapsed += OnTimedEvent;
            this.rand = new Random();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            if (!this.timer.Enabled)
            {
                this.timer.Enabled = true;
                this.timerStarted = DateTime.Now;
                this.button1.Text = "Stop Timer";
            }
            else
            {
                this.timer.Enabled = false;
                this.button1.Text = "Start Timer";
            }
        }

        private void OnTimedEvent(Object source, ElapsedEventArgs e)
        {
            TimeSpan elapsedTime = e.SignalTime - this.timerStarted;
            Invoke(UpdateUI, elapsedTime);
        }

        private void checkBox1_CheckedChanged(object sender, EventArgs e)
        {
            this.isDouble = ((CheckBox)sender).Checked;
        }

        private void UpdateUI(TimeSpan elapsedTime)
        {
            if (elapsedTime.Seconds % 2 == 0)
            {
                this.label1.Text = "The random number is: " + (this.isDouble ? rand.NextDouble().ToString() : rand.Next().ToString());
            }
            this.label2.Text = "Time Elapsed: " + elapsedTime.ToString(@"hh\:mm\:ss");
        }

        private void button2_Click(object sender, EventArgs e)
        {
            new Credits().Show();
        }
    }
}

File: Week2/VB.NET/RandomTimer/RandomTimer/Form1.vb
Imports System.Timers

Public Class Form1
    Private WithEvents Timer As System.Timers.Timer
    Private rnd As Random
    Private isDouble As Boolean = False
    Private timerStarted As DateTime

    Private Sub button1_Click(sender As Object, e As EventArgs) Handles button1.Click
        If Not Me.Timer.Enabled Then
            Me.Timer.Enabled = True
            Me.timerStarted = DateTime.Now
            Me.button1.Text = "Stop Timer"
        Else
            Me.Timer.Enabled = False
            Me.button1.Text = "Start Timer"
        End If
    End Sub

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Me.Timer = New Timers.Timer(1000)
        Me.Timer.AutoReset = True
        Me.rnd = New Random()
    End Sub

    Private Sub OnTimedEvent(sender As Object, e As ElapsedEventArgs) Handles Timer.Elapsed
        Dim elapsedTime As TimeSpan = e.SignalTime - timerStarted
        Invoke(Sub()
                   UpdateUI(elapsedTime)
               End Sub)
    End Sub

    Private Sub UpdateUI(elapsedTime As TimeSpan)
        If elapsedTime.Seconds Mod 2 = 0 Then
            If Me.isDouble Then
                Me.label1.Text = "The random number is: " + Me.rnd.NextDouble().ToString()
            Else
                Me.label1.Text = "The random number is: " + Me.rnd.Next().ToString()
            End If
        End If
        Me.label2.Text = "Time Elapsed: " + elapsedTime.ToString("hh\:mm\:ss")
    End Sub

    Private Sub checkBox1_CheckedChanged(sender As Object, e As EventArgs) Handles checkBox1.CheckedChanged
        Me.isDouble = True
    End Sub

    Private Sub button2_Click(sender As Object, e As EventArgs) Handles button2.Click
        Dim cred As Credits = New Credits()
        cred.Show()
    End Sub
End Class
